File: problems-extend-adapt/0401-0500/0445_backward-digit-sum/solution.go
func sumForwardLinkedDigits(l1 *ListNode, l2 *ListNode) *ListNode {
	// Stacks reverse the reading order without touching the inputs: both
	// least-significant digits end up on top, so the ones columns line up
	// however the lengths differ.
	var stack1 []int
	var stack2 []int
	for node := l1; node != nil; node = node.Next {
		stack1 = append(stack1, node.Val)
	}
	for node := l2; node != nil; node = node.Next {
		stack2 = append(stack2, node.Val)
	}
	// Column addition from the least-significant end. Digits come out
	// least-significant first, so each new node is linked in front of the
	// previous one — front-insertion restores the required
	// most-significant-first order as the loop runs.
	var head *ListNode
	carry := 0
	// One loop condition covers every edge case at once: unequal lengths
	// and a leftover final carry (999 + 1 -> 1000). An empty stack simply
	// contributes nothing.
	for len(stack1) > 0 || len(stack2) > 0 || carry != 0 {
		total := carry
		if len(stack1) > 0 {
			total += stack1[len(stack1)-1]
			stack1 = stack1[:len(stack1)-1]
		}
		if len(stack2) > 0 {
			total += stack2[len(stack2)-1]
			stack2 = stack2[:len(stack2)-1]
		}
		// Split the column total into the new carry and the digit to emit.
		carry = total / 10
		head = &ListNode{Val: total % 10, Next: head}
	}
	return head
}

File: problems-extend-adapt/0401-0500/0445_backward-digit-sum/solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {

    public ListNode sumForwardLinkedDigits(ListNode l1, ListNode l2) {
        // Stacks reverse the reading order without touching the inputs:
        // both least-significant digits end up on top, so the ones
        // columns line up however the lengths differ.
        Deque<Integer> stack1 = new ArrayDeque<>();
        Deque<Integer> stack2 = new ArrayDeque<>();
        while (l1 != null) {
            stack1.push(l1.val);
            l1 = l1.next;
        }
        while (l2 != null) {
            stack2.push(l2.val);
            l2 = l2.next;
        }
        // Column addition from the least-significant end. Digits come out
        // least-significant first, so each new node is linked in front of
        // the previous one — front-insertion restores the required
        // most-significant-first order as the loop runs.
        ListNode head = null;
        int carry = 0;
        // One loop condition covers every edge case at once: unequal
        // lengths and a leftover final carry (999 + 1 -> 1000).
        while (!stack1.isEmpty() || !stack2.isEmpty() || carry != 0) {
            // An empty stack simply contributes nothing.
            int total = carry;
            if (!stack1.isEmpty()) {
                total += stack1.pop();
            }
            if (!stack2.isEmpty()) {
                total += stack2.pop();
            }
            // Split the column total into the new carry and the digit to emit.
            carry = total / 10;
            head = new ListNode(total % 10, head);
        }
        return head;
    }
}
